rust/src/statistics/storage/mod.rs

```diff
@@ -0,0 +1,577 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+use std::{collections::HashSet, fmt, fs::File, io::Write, path::Path};
+
+use crate::{
+    error::BTResult,
+    storage::{self, Checkpointable, DbMode, RootIdProvider, Storage},
+    sync::{
+        Arc, Mutex,
+        atomic::{AtomicBool, Ordering},
+        thread,
+    },
+    types::{ToNodeKind, TreeId},
+};
+
+/// An operation that can be performed on the storage.
+#[derive(Hash, Eq, PartialEq, Debug, Copy, Clone)]
+enum StorageOperation {
+    Get,
+    Set,
+    Reserve,
+    Delete,
+}
+
+impl std::fmt::Display for StorageOperation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StorageOperation::Get => write!(f, "G"),
+            StorageOperation::Set => write!(f, "S"),
+            StorageOperation::Reserve => write!(f, "R"),
+            StorageOperation::Delete => write!(f, "D"),
+        }
+    }
+}
+
+/// A storage operation with associated metadata for logging purposes.k
+#[derive(Hash, Eq, PartialEq, Debug, Clone)]
+struct StorageOperationWithMetadata {
+    op: StorageOperation,
+    timestamp: std::time::Duration,
+    node_kind: String,
+    index: u64,
+}
+
+impl StorageOperationWithMetadata {
+    /// Format the operation with metadata as a CSV line.
+    fn as_csv_line(&self) -> String {
+        format!(
+            "{},{},{},{}",
+            self.op,
+            self.timestamp.as_micros(),
+            &self.node_kind,
+            self.index
+        )
+    }
+}
+
+/// A storage wrapper that logs the performed operations to a file.
+/// It records the type of operation, current timestamp, node kind name and the index of
+/// each operation.
+/// A background worker periodically flushes the collected operations to disk.
+pub struct StorageOperationLogger<S: Storage> {
+    storage: S,
+    operations: Arc<Mutex<HashSet<StorageOperationWithMetadata>>>,
+    first_op_timestamp: Mutex<Option<std::time::Instant>>,
```

Comment:
```suggestion
    start_timestamp: Instant,
```

This seems a bit over-engineered to me: We don't *need* the timestamps to be relative to the first operation. You can just initialize this in `try_new` and get rid of the mutex. In practice this will be a couple of milliseconds difference at most, and if you really wanted everything to be relative to the first operation you can always calculate that during post-processing.

Reply:
Yeah we can simplify this.